File: cs/decryptor.cs
using System;
using System.IO;
using System.Collections.Generic;

class Program
{
    static List FindAllPngData(string filePath)
    {
        List pngDataList = new List();

        using (FileStream file = new FileStream(filePath, FileMode.Open, FileAccess.Read))
        {
            byte[] data = new byte[file.Length];
            file.Read(data, 0, data.Length);

            int startIdx = 0;
            while (true)
            {
                startIdx = Array.IndexOf(data, (byte)0x89, startIdx);
                if (startIdx == -1)
                {
                    break;
                }

                int endIdx = Array.IndexOf(data, (byte)0x82, startIdx + 8); // Finding the end of PNG data marker
                if (endIdx == -1)
                {
                    Console.WriteLine($"End of PNG data not found in the file: {filePath}");
                    return null;
                }

                byte[] pngData = new byte[endIdx - startIdx + 8];
                Array.Copy(data, startIdx, pngData, 0, pngData.Length);
                pngDataList.Add(pngData);

                startIdx = endIdx + 8;
            }
        }

        return pngDataList;
    }

    static void BuildPngFiles(List pngDataList, string outputDir)
    {
        if (!Directory.Exists(outputDir))
        {
            Directory.CreateDirectory(outputDir);
        }

        for (int i = 0; i < pngDataList.Count; i++)
        {
            string outputFilePath = Path.Combine(outputDir, $"output_file_{i + 1}.png");
            using (FileStream outputFile = new FileStream(outputFilePath, FileMode.Create, FileAccess.Write))
            {
                outputFile.Write(pngDataList[i], 0, pngDataList[i].Length);
            }

            Console.WriteLine($"PNG data extracted and saved to: {outputFilePath}");
        }
    }

    static void Main(string[] args)
    {
        string[] filePaths = { "c1", "c2", "c7", "c8", "c9", "mi", "nm1", "nm3", "nm4", "oi" };

        foreach (string inputFilePath in filePaths)
        {
            string outputDir = $"{Path.GetFileNameWithoutExtension(inputFilePath)}_output";
            List pngDataList = FindAllPngData(inputFilePath);
            if (pngDataList != null)
            {
                BuildPngFiles(pngDataList, outputDir);
            }
            else
            {
                Console.WriteLine($"No PNG data found in the file: {inputFilePath}");
            }
        }
    }
}

File: cpp/decryptor.cpp
#include 
#include 
#include 
#include 
#include 

std::vector find_all_png_data(const std::string& file_path) {
    std::ifstream file(file_path, std::ios::binary);
    if (!file.is_open()) {
        std::cerr << "Failed to open file: " << file_path << std::endl;
        return std::vector();
    }

    std::vector data((std::istreambuf_iterator(file)), std::istreambuf_iterator());

    size_t start_idx = 0;
    std::vector> png_data_list;

    while (true) {
        start_idx = data.find("PNG", start_idx);
        if (start_idx == std::string::npos) {
            break;
        }

        size_t end_idx = data.find("IEND\xAE\x42\x60\x82", start_idx + 1);
        if (end_idx == std::string::npos) {
            std::cerr << "End of PNG data not found in the file: " << file_path << std::endl;
            return std::vector();
        }

        std::vector png_data(data.begin() + start_idx, data.begin() + end_idx + 8);
        png_data_list.push_back(png_data);

        start_idx = end_idx + 1;
    }

    return png_data_list;
}

void build_png_files(const std::vector>& png_data_list, const std::string& output_dir) {
    if (mkdir(output_dir.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH) != 0) {
        std::cerr << "Failed to create output directory: " << output_dir << std::endl;
        return;
    }

    for (size_t i = 0; i < png_data_list.size(); ++i) {
        const std::vector& png_data = png_data_list[i];
        std::string output_path = output_dir + "/output_file_" + std::to_string(i + 1) + ".png";
        std::ofstream outfile(output_path, std::ios::binary);
        if (outfile.is_open()) {
            outfile.write(png_data.data(), png_data.size());
            std::cout << "PNG data extracted and saved to: " << output_path << std::endl;
        } else {
            std::cerr << "Failed to write output file: " << output_path << std::endl;
        }
    }
}

int main() {
    std::vector file_paths = {"c1", "c2", "c7", "c8", "c9", "mi", "nm1", "nm3", "nm4", "oi"};

    for (const std::string& input_file_path : file_paths) {
        std::string output_dir = input_file_path.substr(0, input_file_path.find_last_of('.')) + "_output";
        std::vector> png_data_list = find_all_png_data(input_file_path);
        if (!png_data_list.empty()) {
            build_png_files(png_data_list, output_dir);
        } else {
            std::cerr << "No PNG data found in the file: " << input_file_path << std::endl;
        }
    }

    return 0;
}
